Discovery Service RFC

reinkrul opened the pull request:
Things to decide on:
- Do we want to use PEX to determine what credentials to present? Otherwise, implementations will have to hardcode how to build presentations (with which credentials) for each list. It can then also be used to validate presentations to be added to the list. Presentation definitions can be distributed out-of-band, or hosted by the list (e.g. query `/<list-endpoint>/metadata.json`) -> decided have a use case definitions, created by use case maintainers and shared with all users of the use case
- The list is now just that: a list of presentations. It's hard for clients to see which entries changed (were added/removed), which might be useful since checking the presentations might become (more) expensive when the list grows (e.g. entries might be converted to a SQL table for searchability). But recognizing presentations is hard, options: -> decided to let clients query deltas
  - Use presentation.ID, but guaranteeing uniqueness is hard/impossible (since it's set by the presentation creator)
  - Use hash of the presentation (JWT can be hashed as-is, for JSON-LD maybe use proof hash/signature?)

woutslakhorst commented:
- yes we want to use PEX for credential selection. It is distributed by the use case.
- use `jwt_vc` and check the `jti`, `iat` and `exp`. If still valid check DB for `jti`, if not present, check signature.

woutslakhorst commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -0,0 +1,139 @@
+List entries can be a:
+
+- presentation in JWT format as JSON string.
+- presentation in JSON-LD format as JSON object.
```
just one format will be enough for now.

woutslakhorst commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -0,0 +1,139 @@
+### 3.2 Registering on the list
+
+Clients can register a Verifiable Presentation on the list by sending an HTTP POST to the maintainer's list endpoint.
+The HTTP request body MUST be a Verifiable Presentation encoded as JSON. The content type of the request MUST be `application/json`.
```
I would say MUST be a jwt

woutslakhorst commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -0,0 +1,139 @@
+
+The maintainer MUST validate the Verifiable Presentation as specified in section 4. If the validation fails, it MUST return a 400 Bad Request response.
+If the validation succeeds, the Verifiable Presentation MUST be added to the list and the maintainer MUST return a 201 Created response.
+Credential subjects on the list MUST be unique, so if the subject already exists, the new presentation replaces the old one. 
```
definition of unique? presenter and list of VCs?

woutslakhorst commented:
Remove JSON-LD support since the semantic meaning is not used. There's only a check against the PD. This could also allow for forward compatibility. JSON-LD will definitely not be forward compatible.

woutslakhorst reviewed (commented):
Wouldn't it be better to replace `state` with timestamps or a lamport clock?
Timestamps don't require any additional storage but suffer from nanosecond race conditions. A lamport clock doesn't suffer from this, any new POST is just a +1. This could also help with a tombstone list: [~57, 59, 60], up to 57 are removed as well as 59 and 60.

woutslakhorst commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -90,20 +90,24 @@ If one or more presentations are not valid, it SHOULD NOT reject the other prese
+### 3.5. Removing entries
+
+TODO: Clients may remove a credential subject by performing an HTTP DELETE to the list endpoint, where the body contains an empty Verifiable Presentation signed by the credential subject.
+TODO: If we support this: we need to serve a tombstone set as well (and also for revocations)
```
If you add a lamport clock per entry and a tombstone list then you have the feature complete.

woutslakhorst commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -90,20 +90,24 @@ If one or more presentations are not valid, it SHOULD NOT reject the other prese
-- ``issuanceDate`` of the presentation MUST have passed.
-- ``expirationDate`` of the presentation MUST NOT have passed.
-- ``expirationDate`` MUST be after ``issuanceDate``.
+- ``nbf`` (not before) of the presentation MUST have passed.
```
I think iat is usually preferred over nbf.

woutslakhorst commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -128,6 +128,7 @@ Presentations and credentials MUST be in JWT format.
 
 Maintainers MUST share a JSON document describing the list. This document is known as the _list definition_.
 The document MUST contain the following properties:
+- `id` REQUIRED. JSON string containing a value identifies the list. MAY be used to version the definition.
```
call the `JSON strings` just `string`

woutslakhorst reviewed (changes requested):
missing security and privacy considerations.

gerardsn reviewed (changes requested):
Normally the response to a Presentation Definition is a (list of) VP(s) + Presentation Submission. We're missing the PS here.

I got a bit carried away during reviewing, sorry

reinkrul commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -0,0 +1,221 @@
+- `tombstones` REQUIRED. MUST contain a JSON array with zero or more presentation IDs that have been removed from the list.
+- `timestamp` REQUIRED. MUST be a JSON number containing the Lamport timestamp of the last entry.
+
+The `timestamp` query parameter MAY be used by the client to request a delta next time it reads the list.
+If no `timestamp` query parameter is provided, the maintainer MUST return the full list.
```
first available LC, which is also the entire list. Due to the rather short-livedness of presentations the list will stay rather short by nature

reinkrul commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -0,0 +1,221 @@
+The maintainer MUST return a 200 OK response with a JSON object containing:
+- `entries` REQUIRED. MUST contain a JSON array with zero or more presentations.
+- `tombstones` REQUIRED. MUST contain a JSON array with zero or more presentation IDs that have been removed from the list.
+- `timestamp` REQUIRED. MUST be a JSON number containing the Lamport timestamp of the last entry.
```
Could be a implementation suggestion in this RFC

jorritspee commented:
Sorry to enter the dialogue late and without proper preparation. My question is whether you guys see a relation between the use case list and the DCAT-standard or not? DCAT is mentioned often in relation to FAIR-datapoints.

reinkrul commented:
> Sorry to enter the dialogue late and without proper preparation. My question is whether you guys see a relation between the use case list and the DCAT-standard or not? DCAT is mentioned often in relation to FAIR-datapoints.

I had to look up DCAT, but I understand it describes the models of the data sets. The goal of this RFC is to define a way to discover parties (and ultimately services) supporting a certain feature/use case, but not how information/data exchanged within that use case is structured/modeled.

I could imagine use case definitions could include URLs/references to a schema or catalog describing the information model of the use case. On the other hand, the Nuts node does not deal with the actual information itself (which is exchanged directly between the XIS's). But, if the use case definition would be used in other systems (than the Nuts node), it might have some value?

reinkrul commented on rfc/rfc0XX-use-case-list.md:
```diff
@@ -0,0 +1,221 @@
+Clients can remove presentations from a specific credential subject from a list, e.g. when a care organization stops supporting the use case.
+To remove presentations of a credential subject the client MUST send an HTTP DELETE request to the list endpoint,
+where the body contains an empty Verifiable Presentation signed by the credential subject.
+Any credentials in the presentation MUST be ignored by the maintainer.
```
Will replace it with a VP that specifies the ID of the VP being removed.

jorritspee commented:
> > Sorry to enter the dialogue late and without proper preparation. My question is whether you guys see a relation between the use case list and the DCAT-standard or not? DCAT is mentioned often in relation to FAIR-datapoints.
> 
> 
> 
> I had to look up DCAT, but I understand it describes the models of the data sets. The goal of this RFC is to define a way to discover parties (and ultimately services) supporting a certain feature/use case, but not how information/data exchanged within that use case is structured/modeled.
> 
> 
> 
> I could imagine use case definitions could include URLs/references to a schema or catalog describing the information model of the use case. On the other hand, the Nuts node does not deal with the actual information itself (which is exchanged directly between the XIS's). But, if the use case definition would be used in other systems (than the Nuts node), it might have some value?

Hi Rein,
Thanks for your reply. DCAT is generally used by a partner to publish what data sets and/or - services a party provides. I would think that a use case could map to a data service. @hietkamp, you also had some ideas on using DCAT in a Nuts context didn't you?

reinkrul commented:
Renamed to Discovery Service with Client and Server, removed tombstone set in favor of "just" a Verifiable Presentation that refers a retracted Verifiable Presentation.

hietkamp commented:
We plan to agree on a data catalog (dcat2) with KIK-V. This would be a service address in Nuts to an endpoint where the data catalog can be found. The data catalog is intended to provide information about the datasets you offer and in accordance with which standard you offer this data (the ontology in KIK-V).

I myself see more value in mentioning the ontology (you can call it a domain model) that you support. You make data available in accordance with an (version of an) ontology. The principle is that we want to separate data and functionality.

An ontology is suitable for a use case, that's what you agree on. But the ontology can also be used for many other use cases. The use case is also owned by the data user and not the data holder in my experience. Then it is strange that the holder specifies the use cases for which his data is suitable.

The RFC therefore plays into the use case-driven way of working while we want to move towards a data-driven way of working. We may still need this for the first plateau of the NVS, but then we will immediately build an architectural debt towards the second plateau of the NVS.

woutslakhorst commented:
> Then it is strange that the holder specifies the use cases for which his data is suitable.

A lot of the current development for data exchange is use-case oriented and even defines a specific information model per use-case. If all those models converge in the future, there indeed could be a 'single use-case' that only requires a party to publish its data catalog endpoint. This discovery service would then still be useable. It would be a single use-case, single instance usage. A way to discover other endpoints will always be required.

woutslakhorst commented on rfc/rfc0XX-discovery-service.md:
```diff
@@ -0,0 +1,232 @@
+
+### 3.4 Retracting presentations
+
+Clients can remove presentations from the list, e.g. when a care organization stops supporting the use case.
```
RFCs are independent of the domain

gerardsn reviewed (changes requested):
Introduction, security, and privacy considerations are missing.

reinkrul commented on rfc/rfc0XX-discovery-service.md:
```diff
@@ -0,0 +1,232 @@
+### 3.4 Retracting presentations
+
+Clients can remove presentations from the list, e.g. when a care organization stops supporting the use case.
+To remove presentations of a credential subject the client MUST send an HTTP DELETE request to the service endpoint,
```
Note to self: just use the same HTTP POST, only with different body

woutslakhorst commented:
@reinkrul is the name/id of the use case included in the VP? Otherwise you can replay a VP to another use case....

reinkrul commented:
Renamed "timestamp" to "tag", as it is an opaque value. "Timestamp" suggests it's a time-based property, which it might not be.

reinkrul commented:
> @reinkrul is the name/id of the use case included in the VP? Otherwise you can replay a VP to another use case....

This is now included;

> - `aud` MUST contain the ID of the service.

reinkrul commented:
> [JWT-VC profile](https://identity.foundation/jwt-vc-presentation-profile/#nbf-jwt-claim)
>
>[VC Data Model v1.1](https://identity.foundation/jwt-vc-presentation-profile/#term:vc-data-model-v1.1) specifies that “issuanceDate” property MUST be represented as an nbf JWT claim, and not iat JWT claim. This might sound couterintuitive, but the implementers of this profile MUST follow this guidance.
>
>Do we need a similar statement? (If only to prevent reconsidering/discussions in the future)

If we already refer to VC 1.1, it's covered? (Since it's a MUST)

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+The server MUST validate the Verifiable Presentation as specified in section 4. If the validation fails, it MUST return a 400 Bad Request response.
+If the validation succeeds, the Verifiable Presentation MUST be added to the list and the server MUST return a 201 Created response.
+
+A credential subject (identified by `credentialSubject.id`) MUST NOT appear more than once on the list,
+so a new registration MUST replace the previous one from the same credential subject.
```
`the list` probably refers to §3.2. This is confusing for the uninitiated

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+Clients MUST read the list by sending an HTTP GET to the server's service endpoint.
+The server MUST return a 200 OK response with a JSON object containing:
+- `entries` REQUIRED. MUST contain a JSON array with zero or more presentations.
+- `tag` REQUIRED. MUST be a JSON string referring to the last entry. The tag is an opaque value; the client SHOULD NOT derive any meaning from it.
```
> referring to the last entry.

It refers to the moment in time of the last entry (in the ref impl). A description like: "used by the server to provide delta's" might be better suited

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+``
+
+Clients MUST validate each presentation in the list as specified in section 4.
+If a presentation is valid, the client uses it in its system. If a presentation is not valid, it MUST be rejected.
```
```suggestion
If a presentation is not valid, it MUST be rejected.
```

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+If a presentation is valid, the client uses it in its system. If a presentation is not valid, it MUST be rejected.
+If one or more presentations are not valid, it SHOULD NOT reject the other presentations.
+
+### 3.3 List pruning
```
How does this work with expired (not retracted) VPs? Do clients also have to prune their lists?

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+- `exp` MUST be after `nbf`.
+- `aud` MUST contain the ID of the service.
+- the number of seconds between `nbf` and `exp` MUST NOT exceed `presentation_max_validity` (see Service Definition).
+- all credential issuers MUST be trusted (see section 5). 
```
implicit through Presentation Definition?

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+
+## 6. Trust
+
+Trust of credential issuers (e.g. `did:example:education-accredetor` issuing `EducationalInstitutionCredential`) SHOULD be defined by the presentation definition.
```
`accreditor`

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+    "path": "$.issuer",
+    "filter": {
+      "type": "string",
+      "const": "did:example:education-accredetor"
```
`accreditor`

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+  }
+]
+``
+
```
Missing security and privacy considerations

woutslakhorst commented on SUMMARY.md:
```diff
@@ -20,6 +20,7 @@
 * [RFC019 Employee Identity Authentication Means](rfc/rfc019-employee-identity-means.md)
 * [RFC020 Authorization credential extension](rfc/rfc020-authorization-credential-extension.md)
 * [RFC021 VP Token Grant Type](rfc/rfc021-vp_token-grant-type.md)
+* [RFC022 Discovery Service](rfc/rfc022-discovery-service.md)
```
Also update graph image

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,252 @@
+
+* [JWT] M. Jones, J. Bradley, N. Sakimura, "JSON Web Token (JWT)", RFC 7519, May 2015, <https://www.rfc-editor.org/info/rfc7519>.
+* [VC-DATA-MODEL] M. Sporny, D. Longley, D. Chadwick, "Verifiable Credentials Data Model 1.1", W3C Recommendation, 3 March 2022, <https://www.w3.org/TR/vc-data-model/>.
+* [PE] D. Buchner, B. Zundel, M. Riedel, K.H. Duffy, "Presentation Exchange 2.0.0", 12 September 2023, <https://identity.foundation/presentation-exchange/spec/v2.0.0/>.
```
This is usually referred to as `PEX` since that sounds a lot nicer when pronounced.

gerardsn reviewed (changes requested):
A credential is not the same as a verifiable credential. The same goes for presentation/VP. I'd prefer to add VP and VC to the terminology section and replace all.

woutslakhorst reviewed (changes requested):
missing error return

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+
+If a validation step fails, the presentation MUST be rejected.
+
+JWT credential and presentation encoding as specified by [VC-DATA-MODEL] MUST be applied.
```
is this saying that credentials also MUST be JWT?

woutslakhorst replied:
should only be presentation

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+The optional `DiscoveryRegistrationCredential` is used to add additional parameters to the registration.
+It's a _holder_ credential, meaning it has been self issued. Fields under `credentialSubject` (except `id`) can be used by clients depending on the use case.
+The credential has the following requirements:
+- `type` MUST be `DiscoveryRegistrationCredential`.
```
```suggestion
- `type` MUST contain `DiscoveryRegistrationCredential`.
```

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+JWT credential and presentation encoding as specified by [VC-DATA-MODEL] MUST be applied.
+A clock skew of 5 seconds MAY be applied to `nbf` and `exp` claims.
+
+### 4.1 DiscoveryRegistrationCredential
```
add something about missing a proof / sharing proof with the VP

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+### 4.1 DiscoveryRegistrationCredential
+
+The optional `DiscoveryRegistrationCredential` is used to add additional parameters to the registration.
+It's a _holder_ credential, meaning it has been self issued. Fields under `credentialSubject` (except `id`) can be used by clients depending on the use case.
```
reference `holder credential` concept (Has it been removed from VC data model v2? I can't find it.)

woutslakhorst replied:
`holder claims`: https://www.w3.org/TR/vc-data-model-2.0/#presentations-including-holder-claims

woutslakhorst replied:
changed to `self-asserted`

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+  the JWT `kid` header MUST reference an `assertionMethod` key from the subject's DID document.
+- The DID of the credential subject MUST be of a DID method defined in the service definition.
+- the presentation MAY contain a `DiscoveryRegistrationCredential` credential defined by the Nuts JSON-LD context (https://nuts.nl/credentials/v1).
+  If present, it MUST NOT be included in the next validation steps.
```
what if it is included in the presentation definition (next validation step)?

It makes sense to me that a use-case lists all the endpoints in the PD that are required for the use-case

woutslakhorst replied:
added an unless.

gerardsn replied:
I don't like magic and this also makes it harder to implement. It's unlikely a use-case will only contain the `authServerURL`, so I prefer the credential to be defined explicitly. If needed we can also add a flag to the API to add the `authServerURL` that defaults to true.

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+
+## 7.1 Denial of Service
+
+Since the server is hosting a public endpoint, it should take the following measures to prevent denial-of-service attacks:
```
Should this apply to all nodes?
All nodes have the endpoint exposed and forward to what they consider the server. (Implementation detail of the nuts-node, so may need to be added to documentation on the node if valid.)

woutslakhorst replied:
according to the spec, clients don't expose anything.

gerardsn commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+
+## 7.2 Trust
+
+Designers of services should build trust into the Presentation Definition of the server. This can be done by specifying specific issuers of credentials.
```
```suggestion
Designers of services should write trust into the Presentation Definition of the Service Definition. This can be done by specifying specific issuers of credentials.
```

gerardsn reviewed (changes requested):
add [seed](https://github.com/nuts-foundation/nuts-node/issues/3463)

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+### 3.1 Registration
+
+Clients can publish a Verifiable Presentation to the server by sending it in an HTTP POST to the service endpoint.
+The HTTP request body MUST be a Verifiable Presentation (in JWT format). The content type of the request MUST be `application/json`.
```
we could drop this

woutslakhorst commented on rfc/rfc022-discovery-service.md:
```diff
@@ -0,0 +1,299 @@
+Clients processing a retraction entry MUST remove the presentation indicated by `retract_jti`.
+The expiration of the retraction presentation MUST be equal to the expiration of the presentation to retract.
+
+### 3.5 Error responses
```
why specify the exact error string? There's nothing a user can do. All these errors are caused by a faulty client.

gerardsn replied:
IMO these are 3 different types of errors
- faulty client. action: nothing, or complain to your software provider (make issue for nuts-node)
- misconfigured client. action: users need to figure out who is at fault -> server/client (probably client)
- server dependency failure. action: monitor the issue and take action if it takes too long

The exact error message could be clearer, but I think these categories would help users debug any issues.

gerardsn replied:
probably needs a general `ServerError` too

woutslakhorst replied:
in all cases an admin has to do something, not a user.

gerardsn replied:
yes, but the error helps the admin figure out what to do. 

However, these suggestions were the result of o the original issue title in https://github.com/nuts-foundation/nuts-node/issues/3471. The changed scope of that issue makes this irrelevant
